// wenay-common/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadCandles = exports.Params = exports.List = exports.Math = exports.ListNodeAnd = exports.Color = exports.Time = exports.Common = exports.BaseTypes = exports.Wait = void 0;
exports.test = test;
__exportStar(require("./Common/node_console"), exports);
__exportStar(require("./Common/Decorator"), exports);
__exportStar(require("./Common/BaseTypes"), exports);
__exportStar(require("./Common/ByteStream"), exports);
__exportStar(require("./Common/Color"), exports);
__exportStar(require("./Common/Time"), exports);
__exportStar(require("./Common/common"), exports);
__exportStar(require("./Common/commonsServer"), exports);
__exportStar(require("./Common/commonsServerMini"), exports);
__exportStar(require("./Common/event"), exports);
__exportStar(require("./Common/funcTimeWait"), exports);
__exportStar(require("./Common/List"), exports);
__exportStar(require("./Common/ListNodeAnd"), exports);
__exportStar(require("./Common/Math"), exports);
__exportStar(require("./Common/MemoFunc"), exports);
__exportStar(require("./Common/objectPath"), exports);
__exportStar(require("./Common/waitRun"), exports);
__exportStar(require("./Common/Listen"), exports);
__exportStar(require("./toError/myThrow"), exports);
__exportStar(require("./Exchange/Bars"), exports);
__exportStar(require("./Exchange/LoadBase"), exports);
__exportStar(require("./Exchange/ConnectData/Binance"), exports);
__exportStar(require("./Common/SocketServerHook"), exports);
exports.Wait = __importStar(require("./Exchange/Bars"));
exports.BaseTypes = __importStar(require("./Common/BaseTypes"));
exports.Common = __importStar(require("./Common/common"));
exports.Time = __importStar(require("./Common/Time"));
exports.Color = __importStar(require("./Common/Color"));
exports.ListNodeAnd = __importStar(require("./Common/ListNodeAnd"));
exports.Math = __importStar(require("./Common/Math"));
exports.List = __importStar(require("./Common/List"));
exports.Params = __importStar(require("./Exchange/CParams"));
exports.LoadCandles = __importStar(require("./Exchange/ConnectData/Binance"));
const ListNodeAnd_1 = require("./Common/ListNodeAnd");
function test() {
    const tt = new ListNodeAnd_1.CListNodeAnd();
    console.log("test");
}
